### ISSUE TITLE
`clippy::no_effect` trips when instantiating an uninhabited type

### ISSUE THREAD

**Issue by PatchMixolydic** [rust-lang/rust-clippy#8033]
Lint name: `clippy::no_effect`

Continuing [the trend of *technically*-false positives that almost nobody will encounter](https://github.com/rust-lang/rust-clippy/issues/7792), I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f2103ed4bfc6ad99cf85bfc2601ff1ea)):

```rust
fn main() {
    unsafe {
        *(1 as *mut std::convert::Infallible);
    }
}
```

Running `cargo clippy` produces the following diagnostic:
```
warning: statement with no effect
 --> src/main.rs:3:9
  |
3 |         *(1 as *mut std::convert::Infallible);
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::no_effect)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#no_effect
```

However, [`Infallible` is uninhabited](https://github.com/rust-lang/rust/blob/862962b90e59c5c1e217df74de80d3a81eee42f4/library/core/src/convert/mod.rs#L683), which seems to indicate that [this statement causes undefined behaviour](https://doc.rust-lang.org/nomicon/transmutes.html), which is most definitely an effect (and a very unpleasant one at that).
```
$ cargo miri run
   Compiling clippy-issue v0.1.0 (/tmp/clippy-issue)
    Finished dev [unoptimized + debuginfo] target(s) in 0.24s
     Running `/home/patchmixolydic/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/cargo-miri target/miri/x86_64-unknown-linux-gnu/debug/clippy-issue`
error: Undefined Behavior: type validation failed: encountered a value of uninhabited type std::convert::Infallible
 --> src/main.rs:3:9
  |
3 |         *(1 as *mut std::convert::Infallible);
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered a value of uninhabited type std::convert::Infallible
  |
  = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
  = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
          
  = note: inside `main` at src/main.rs:3:9
...
```

This *may* cause confusion if someone unknowingly tries to construct an uninhabited type, forgets to assign it to anything, and takes the Clippy warning to mean that it doesn't cause undefined behaviour, but this is astronomically unlikely. Anyone who writes code like this is likely knowingly trying to invoke UB. I doubt this is worth fixing, but just in case...

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.58.0-nightly (8b09ba6a5 2021-11-09)
binary: rustc
commit-hash: 8b09ba6a5d5c644fe0f1c27c7f9c80b334241707
commit-date: 2021-11-09
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```

Also reproduced on `1.58.0-nightly (2021-11-24 b426445c60b4faab6e96)`.

**Comment by llogiq** [comment]
I don't think that UB should be seen as an "effect".

I guess the right thing to do here is to add a lint against instantiating uninhabitable types?

**Comment by PatchMixolydic** [comment]
That's probably the most sensible solution, since it could help people who may have forgotten/don't know the rules around uninhabited types. Such a lint could also be useful for generic code in rare cases, such as in this contrived generic function that can create any ZST ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e58f9b0b7301e31514b778cb352bb8b2)):
```rust
#![feature(generic_const_exprs)]

// ...

fn get_zst<T: Copy>() -> T
where
    Predicate<{ size_of::<T>() == 0 }>: True,
{
    // SAFETY: any aligned pointer is a valid pointer to a ZST
    // (I believe? this is what I've seen in unsafe code, and miri seems to agree)
    // ...unless, of course, that ZST is uninhabited...
    unsafe {
        *(align_of::<T>() as *mut T)
    }
}
```

Of course, designing such a lint might be difficult.

**Comment by llogiq** [comment]
I think I'm going to try to create such a lint.
